docs/proposals/ktl.md

```diff
@@ -0,0 +1,214 @@
+- vendor in kubernetes/kubectl/cmd cobra commands, and register them under `ktl`
+- vendor in commands from other sources as needed over time
+
+Dynamic dispatch:
+
+- support git-style plugins that allow registering new commands under `ktl`
+  - use simplified version of the kubectl plugin implementation - make configuration files optional
+- plugins only purpose is discovery of kubernetes related commands
+  - plugins can leverage shared client cli libraries (whether they are installed as plugins or not)
+- by default throw an error if plugin names conflict with existing commands
+  - this is configurable
+- plugins can be disabled through modifying `~/.ktlconfig`
+
+Overriding existing commands:
+
+- support command alias' to allow overriding one command with another
+  - allows plugins to extend and override built-in commands
+  - allows `ktl kube *` commands to be aliased to `ktl *`
+
+#### Configuration
+
+Use [viper](https://github.com/spf13/viper) to configure dispatch precedence
+
+### Cli commands
+
+All new cli commands should be built outside of kubernetes/kubernetes and vendored into the kubernetes/ktl.  Command repos
+implement an interface that returns a list of cobra (sp13) commands, which ktl can register under the root.
+
+Initially new cli commands can be built in the kubernetes/kubectl, but if needed, development maybe moved to other repos.
+
+#### Conventions
+
+- ktl top level commands are one of
+  - command groups that work against a specific API group/versions (e.g. isto, svcat, kube)  e.g. `create`
+    for a given resource.
+  - generic commands that are agnostic to specific APIs, but instead discover API metadata and work against
+    all APIs following published conventions.  e.g. a `scale` that works against anything with a scale sub resource
+
+
+### Library infrastructure
+
+Develop shared set of client / cli libraries that handle registering flags / environment variables / config files
+and provide common functionality to cli clients of Kubernetes.
+
+- reading and parsing kubeconfig
+- printing libraries
+- reading and parsing config into objects from files, directories, stdin
+- indexing + querying discovery API and openapi data
+- manipulating unstructured objects using openapi schema
+- merging objects
+- generating patches from objects
+- writing objects to the server - optimistic locking failure retries with exp backoff
+- providing tty for subresources that accept - e.g. exec, attach
+- defining shared exit codes (e.g. exit code when not doing a conditional write)
+- test infrastructure for unit and integration tests
+  - support version skew testing
+
+### Build and release infrastructure
+
+Develop build triggers to automatically cut and publish builds based on the presence of GitHub tags.  Aggregate
+release notes from vendored commands.
+
+- use GCP container builder + mirrored GCP source repo
+- publish binary to gs:// bucket
+- publish binary + release notes to GitHub releases page
```

Comment:
Publishing container image would be also very handy, if we use GCP builder, it seems like a no-brainer.

Reply:
yes, agreed